### PR TITLE
batcheval: add MVCC-compliant `RevertRange` variant

### LINKED ISSUE #70416: kvserver: MVCC-compliant `RevertRange` variant
As described in #69380, we need an MVCC-compliant variant of `RevertRange` that allows us to undo recent changes to a key span, typically in order to cancel bulk operations such as `IMPORT INTO`. Conceptually, this needs to iterate across the visible keys at the target revert time (including tombstones), and any keys that have a different most-recent value must be replaced by its original value. In the case of deletes, if there are long runs of keys to delete, it should drop an MVCC range tombstone instead.

~~To accomplish this, we should add a lower timestamp predicate to `DeleteRange` that will only drop tombstones over keys that have a timestamp at or later than the given lower timestamp. It should use range tombstones where possible, to delete long runs of matching keys, but fall back to point tombstones when necessary.~~

See also the (currently internal) [design document](https://docs.google.com/document/d/1ItxpitNwuaEnwv95RJORLCGuOczuS2y_GoM2ckJCnFs/edit#).

Epic CRDB-2624

Jira issue: CRDB-10066

### PR DESCRIPTION
**Pull request by erikgrinaker**
**roachpb: add `isAlone` for `RevertRangeRequest`**

Since `RevertRange` mutates MVCC history, we want them to be alone in a
batch. The DistSender will split any batches that have multiple such
requests.

Release note: None

**storage: add `NextKeyIgnoringTime()` for `MVCCIncrementalIterator`**

This patch adds a method `NextKeyIgnoringTime()` for
`MVCCIncrementalIterator`. This can be used to find the next key (as
opposed to version) of the iterator, ignoring the time bounds. It's
similar to `NextIgnoringTime()`, but calls `NextKey()` instead of
`Next()` on the underlying iterator.

Release note: None

**batcheval: add MVCC-compliant `RevertRange` variant**

This adds a new parameter `ExperimentalPreserveHistory` which, rather
than clearing keys above the target time, will write new values or
tombstones that reflect the state at the target time. For long runs of
new keys, this will instead drop an MVCC range tombstone. This makes the
command respect e.g. MVCC immutability, the closed timestamp, and
timestamp cache.

Note that MVCC range tombstones are currently experimental, and as such
this parameter is also experimental. Callers must call
`storage.CanUseExperimentalMVCCRangeTombstones()` before using it.

Resolves #70416.

Release note: None

### CODE REVIEW

File: pkg/roachpb/api.go [hunk 1]
```diff
@@ -1263,7 +1263,13 @@ func (*ClearRangeRequest) flags() flag { return isWrite | isRange | isAlone }
 
 // Note that RevertRange commands cannot be part of a transaction as
 // they clear all MVCC versions above their target time.
-func (*RevertRangeRequest) flags() flag { return isWrite | isRange }
+func (rrr *RevertRangeRequest) flags() flag {
+	if rrr.ExperimentalPreserveHistory {
+		return isRead | isWrite | isRange | isAlone | updatesTSCache | appliesTSCache
+	}
+	// TODO(erikgrinaker): add isAlone?
```
Comment (dt) [inline pkg/roachpb/api.go:1270]: I think so, yes

File: pkg/roachpb/api.proto [hunk 2]
```diff
@@ -402,18 +402,34 @@ message ClearRangeResponse {
...
 
-  // TargetTime specifies a the time to which to "revert" the range by clearing
-  // any MVCC key with a strictly higher timestamp. TargetTime must be higher
+  // TargetTime specifies a the time to which to "revert" the range to. Any
+  // versions later than TargetTime will be undone. TargetTime must be higher
   // than the GC Threshold for the replica - so that it is assured that the keys
   // for that time are still there — or the request will fail.
   util.hlc.Timestamp target_time = 2 [(gogoproto.nullable) = false];
 
+  // ExperimentalPreserveHistory will preserve MVCC history by, rather than
```
Comment (dt) [inline pkg/roachpb/api.proto:419]: We _could_ decouple these, and say 1) PreserveHistory will do the tombstone/undo, and is thus mvcc and 2) experimental_mvcc_delrange allows preserve history, if in use, to use cheap range deletion tombstones rather than slow/expensive individual tombstones. Then we can do mvcc but just slow without being in experimental territory?

File: pkg/storage/mvcc_incremental_iterator.go [hunk 3]
```diff
@@ -533,6 +533,33 @@ func (i *MVCCIncrementalIterator) NextIgnoringTime() {
 	}
 }
 
+// NextKeyIgnoringTime returns the next distinct key that would be encountered
```
Comment (dt) [inline pkg/storage/mvcc_incremental_iterator.go:536]: should this be in one of the prior commits? I guess waiting to add it here gets it test coverage.

**Review by dt — COMMENTED** [review]
Just one API question from me

File: pkg/roachpb/api.proto (hunk 2 above)
Comment (erikgrinaker) [reply pkg/roachpb/api.proto:419]: We could, but that would require more testing and such, and I can't afford that right now. Any reason to believe we would use this in 22.1?

File: pkg/storage/mvcc_incremental_iterator.go (hunk 3 above)
Comment (erikgrinaker) [reply pkg/storage/mvcc_incremental_iterator.go:536]: I could split it out and add some unit tests. That's probably a good idea anyway, since this could get picked up for production usage somewhere.

File: pkg/roachpb/api.proto (hunk 2 above)
Comment (dt) [reply pkg/roachpb/api.proto:419]: nope, works for me.

File: pkg/storage/mvcc.go [hunk 4]
```diff
@@ -2218,6 +2274,211 @@ func recordIteratorStats(traceSpan *tracing.Span, iteratorStats IteratorStats) {
...
+					int64(encodedMVCCKeyLength(MVCCKey{Key: deleteBuf[0], Timestamp: writeTimestamp}))
+				deleteRangeStart = deleteBuf[0]
+				deleteBufIdx = 0
+				deleteBufBytes = 0
+			}
+		}
+		return nil, nil
+	}
+
+	// We set up an ingremental iterator from the revert time to look for any
```
Comment (aliher1911) [inline pkg/storage/mvcc.go:2398]: incremental

File: pkg/storage/mvcc.go [hunk 5]
```diff
@@ -2218,6 +2274,211 @@ func recordIteratorStats(traceSpan *tracing.Span, iteratorStats IteratorStats) {
...
+		EndKey:                              endKey,
+		StartTime:                           revertTimestamp,
+		EndTime:                             writeTimestamp, // puts will error on any newer versions
+	})
+	defer iter.Close()
+
+	var revertKey roachpb.Key
+	var revertValue, revertValueFrom []byte
+	iter.SeekGE(MVCCKey{Key: startKey})
+	for {
```
Comment (aliher1911) [inline pkg/storage/mvcc.go:2414]: I found following mvcc iterator loops where we iterate versions and keys and feed it to "state machine" pretty hard to follow. Maybe the better idea would be to nest iteration of versions into iteration of keys and only track deletions inbetween?
We may leave it as a todo since we have some deadlines.

**Review by aliher1911 — COMMENTED** [review]
Mostly nits about comments, but I think there's some room for improvement in ExperimentalMVCCRevertRange which we should address before in becomes non-experimental.

File: pkg/storage/mvcc_incremental_iterator.go [hunk 6]
```diff
@@ -518,6 +533,34 @@ func (i *MVCCIncrementalIterator) NextIgnoringTime() {
...
+// in a non-incremental iteration by moving the underlying non-TBI iterator
```
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply pkg/storage/mvcc_incremental_iterator.go:537]: Updated.

**Review by dt — APPROVED** [review]
API as it'll be available to bulk/sql clients LGTM!

**Comment by erikgrinaker** [comment]
TFTRs!

bors r=aliher1911,dt

**Comment by erikgrinaker** [comment]
Some weird agent failure for `git checkout`.

bors retry

**Comment by erikgrinaker** [comment]
Unrelated flake. Third time's the charm!

bors retry
